File: python/agent-protocols/src/agent_protocols/identity.py
from __future__ import annotations

import base64
import hashlib
import json
import re
import time
from dataclasses import dataclass
from typing import Any, MutableMapping, Protocol

import rfc8785
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from .errors import AgentProtocolError

AGENT_ID_PREFIX = "did:agent:"
DEFAULT_LIVE_WRITE_WINDOW_MS = 300_000
DEFAULT_NONCE_TTL_MS = 300_000
DEFAULT_REQUEST_JWT_TTL_SECS = 300
MAX_NONCE_HEADER = "Max-Seen-Nonce"
MAX_SAFE_NONCE = 0x1FFFFFFFFFFFFF

Event = dict[str, Any]
Envelope = dict[str, Any]
AgentId = str


def agent_id_from_public_key(public_key: bytes) -> AgentId:
    if len(public_key) != 32:
        raise AgentProtocolError("invalid_public_key", f"public key must be 32 bytes, got {len(public_key)}")
    return f"{AGENT_ID_PREFIX}{_base64url_encode(public_key)}"


def public_key_bytes(agent_id: AgentId) -> bytes:
    if not agent_id.startswith(AGENT_ID_PREFIX):
        raise AgentProtocolError("invalid_agent_id", "agent id must start with did:agent:")
    data = _base64url_decode_no_pad(agent_id[len(AGENT_ID_PREFIX):])
    if len(data) != 32:
        raise AgentProtocolError("invalid_public_key", f"agent id public key must be 32 bytes, got {len(data)}")
    return data


def validate_agent_id(agent_id: AgentId) -> AgentId:
    public_key_bytes(agent_id)
    return agent_id


@dataclass(frozen=True)
class RequestBinding:
    audience: str

    @classmethod
    def create(cls, audience: str) -> "RequestBinding":
        return cls(audience=audience)


class AgentSigner:
    def __init__(self, private_key: Ed25519PrivateKey):
        self._private_key = private_key

    @classmethod
    def generate(cls) -> "AgentSigner":
        return cls(Ed25519PrivateKey.generate())

    @classmethod
    def from_seed(cls, seed: bytes) -> "AgentSigner":
        if len(seed) != 32:
            raise AgentProtocolError("invalid_private_key", f"seed must be 32 bytes, got {len(seed)}")
        return cls(Ed25519PrivateKey.from_private_bytes(seed))

    def public_key(self) -> bytes:
        return self._private_key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)

    def agent_id(self) -> AgentId:
        return agent_id_from_public_key(self.public_key())

    def sign_event(self, event: Event) -> Envelope:
        digest = event_hash_bytes(event)
        return {
            "hash": _base64url_encode(digest),
            "event": event,
            "signature": sign_event_hash(self._private_key, digest),
        }

    def sign_request_jwt(self, claims: dict[str, Any]) -> str:
        agent_id = self.agent_id()
        if claims.get("iss") != agent_id or claims.get("sub") != agent_id:
            raise AgentProtocolError("invalid_jwt_claim", "iss and sub must match the signing agent id")
        header = {"alg": "EdDSA", "typ": "JWT", "kid": agent_id}
        encoded_header = _base64url_encode(json.dumps(header, separators=(",", ":")).encode())
        encoded_payload = _base64url_encode(json.dumps(claims, separators=(",", ":")).encode())
        signing_input = f"{encoded_header}.{encoded_payload}".encode()
        signature = self._private_key.sign(signing_input)
        return f"{encoded_header}.{encoded_payload}.{_base64url_encode(signature)}"


class NonceStore(Protocol):
    def check_and_update(self, actor: AgentId, nonce: int, now_ms: int, ttl_ms: int) -> int: ...

    def max_nonce(self, actor: AgentId, now_ms: int) -> int | None: ...


class MemoryNonceStore:
    def __init__(self) -> None:
        self._records: dict[AgentId, tuple[int, int]] = {}

    def check_and_update(self, actor: AgentId, nonce: int, now_ms: int, ttl_ms: int) -> int:
        validate_nonce(nonce)
        if ttl_ms < 0:
            raise AgentProtocolError("invalid_nonce", "nonce cache ttl must be non-negative")
        record = self._records.get(actor)
        if record is not None:
            max_nonce, expires_at = record
            if expires_at > now_ms and nonce <= max_nonce:
                raise AgentProtocolError("nonce_not_greater", f"nonce must be greater than accepted max nonce {max_nonce}")
        self._records[actor] = (nonce, now_ms + ttl_ms)
        return nonce

    def max_nonce(self, actor: AgentId, now_ms: int) -> int | None:
        record = self._records.get(actor)
        if record is None:
            return None
        max_nonce, expires_at = record
        return max_nonce if expires_at > now_ms else None


class ClientNonceManager:
    def __init__(self, next_nonce: int = 1) -> None:
        validate_nonce(next_nonce)
        self._next_nonce = next_nonce

    def peek(self) -> int:
        return self._next_nonce

    def next_nonce(self) -> int:
        nonce = self._next_nonce
        validate_nonce(nonce)
        self._next_nonce += 1
        return nonce

    def observe_max_nonce(self, max_nonce: int | str | None) -> None:
        if max_nonce is None or max_nonce == "":
            return
        try:
            parsed = int(max_nonce)
        except (TypeError, ValueError) as exc:
            raise AgentProtocolError("invalid_nonce", "invalid max nonce header") from exc
        validate_nonce(parsed)
        if parsed >= self._next_nonce:
            self._next_nonce = parsed + 1


def create_event(protocol: str, event_type: str, actor: AgentId, created_at: int, nonce: int, payload: Any) -> Event:
    validate_agent_id(actor)
    validate_nonce(nonce)
    return {
        "protocol": protocol,
        "type": event_type,
        "actor": actor,
        "created_at": created_at,
        "nonce": nonce,
        "payload": payload,
    }


def with_room_id(event: Event, room_id: str) -> Event:
    next_event = dict(event)
    next_event["room_id"] = room_id
    return next_event


def canonical_event_bytes(event: Event) -> bytes:
    canonical = rfc8785.dumps(event)
    return canonical if isinstance(canonical, bytes) else canonical.encode()


def event_hash(event: Event) -> str:
    return _base64url_encode(event_hash_bytes(event))


def event_hash_bytes(event: Event) -> bytes:
    validate_nonce(event["nonce"])
    return hashlib.sha3_256(canonical_event_bytes(event)).digest()


def sign_event(private_key: Ed25519PrivateKey, event: Event) -> str:
    return sign_event_hash(private_key, event_hash_bytes(event))


def sign_event_hash(private_key: Ed25519PrivateKey, event_hash: bytes) -> str:
    return _base64url_encode(private_key.sign(_valid_event_hash_bytes(event_hash)))


def verify_event_hash(envelope: Envelope) -> None:
    expected = event_hash(envelope["event"])
    actual = envelope["hash"]
    if expected != actual:
        raise AgentProtocolError("invalid_event_hash", f"invalid event hash: expected {expected}, got {actual}")


def verify_signature(envelope: Envelope) -> None:
    public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes(envelope["event"]["actor"]))
    verify_event_hash_signature(public_key, event_hash_bytes(envelope["event"]), envelope["signature"])


def verify_event_hash_signature(public_key: Ed25519PublicKey, event_hash: bytes, encoded_signature: str) -> None:
    signature = _base64url_decode(encoded_signature)
    if len(signature) != 64:
        raise AgentProtocolError("invalid_signature", f"signature must be 64 bytes, got {len(signature)}")
    try:
        public_key.verify(signature, _valid_event_hash_bytes(event_hash))
    except InvalidSignature as exc:
        raise AgentProtocolError("invalid_signature", "signature verification failed") from exc


def verify_envelope(envelope: Envelope) -> None:
    verify_event_hash(envelope)
    verify_signature(envelope)


def verify_timestamp(created_at: int, now_ms: int, window_ms: int) -> None:
    if window_ms < 0 or abs(created_at - now_ms) > window_ms:
        raise AgentProtocolError("timestamp_out_of_window", "timestamp is outside the allowed live-write window")


def verify_live_envelope(envelope: Envelope, nonce_store: NonceStore, *, now_ms: int | None = None, window_ms: int = DEFAULT_LIVE_WRITE_WINDOW_MS, nonce_ttl_ms: int = DEFAULT_NONCE_TTL_MS) -> int:
    current_now_ms = now_ms if now_ms is not None else unix_ms()
    verify_envelope(envelope)
    verify_timestamp(envelope["event"]["created_at"], current_now_ms, window_ms)
    return nonce_store.check_and_update(envelope["event"]["actor"], envelope["event"]["nonce"], current_now_ms, nonce_ttl_ms)


def create_request_jwt_claims(agent_id: AgentId, binding: RequestBinding, issued_at: int, ttl_secs: int) -> dict[str, Any]:
    return {
        "iss": agent_id,
        "sub": agent_id,
        "aud": binding.audience,
        "iat": issued_at,
        "exp": issued_at + ttl_secs,
    }


def verify_request_jwt(token: str, *, audience: str, now_secs: int | None = None, max_ttl_secs: int = DEFAULT_REQUEST_JWT_TTL_SECS) -> dict[str, Any]:
    parts = token.split(".")
    if len(parts) != 3:
        raise AgentProtocolError("invalid_jwt", "expected three compact JWS parts")
    header = json.loads(_base64url_decode(parts[0]))
    claims = json.loads(_base64url_decode(parts[1]))
    signature = _base64url_decode(parts[2])
    signing_input = f"{parts[0]}.{parts[1]}".encode()

    if header.get("alg") != "EdDSA":
        raise AgentProtocolError("invalid_jwt_claim", "alg must be EdDSA")
    if header.get("typ") != "JWT":
        raise AgentProtocolError("invalid_jwt_claim", "typ must be JWT")
    if header.get("kid") != claims.get("iss") or claims.get("iss") != claims.get("sub"):
        raise AgentProtocolError("invalid_jwt_claim", "kid, iss, and sub must identify the same Agent ID")

    public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes(header["kid"]))
    try:
        public_key.verify(signature, signing_input)
    except InvalidSignature as exc:
        raise AgentProtocolError("invalid_signature", "JWT signature verification failed") from exc

    if claims.get("aud") != audience:
        raise AgentProtocolError("invalid_jwt_claim", "aud mismatch")

    now = now_secs if now_secs is not None else unix_secs()
    if claims["exp"] <= claims["iat"]:
        raise AgentProtocolError("invalid_jwt_claim", "exp must be greater than iat")
    if claims["iat"] > now or claims["exp"] < now:
        raise AgentProtocolError("invalid_jwt_claim", "iat/exp outside valid time window")
    if claims["exp"] - claims["iat"] > max_ttl_secs:
        raise AgentProtocolError("invalid_jwt_claim", "JWT ttl exceeds maximum")
    return claims


def unix_ms() -> int:
    return int(time.time() * 1000)


def unix_secs() -> int:
    return int(time.time())


def validate_nonce(nonce: int) -> None:
    if not isinstance(nonce, int) or nonce < 1 or nonce > MAX_SAFE_NONCE:
        raise AgentProtocolError("invalid_nonce", "nonce must be a positive integer less than or equal to 9007199254740991")


def _base64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _base64url_decode(value: str) -> bytes:
    padding = "=" * ((4 - len(value) % 4) % 4)
    return base64.urlsafe_b64decode(value + padding)


def _base64url_decode_no_pad(value: str) -> bytes:
    if not re.fullmatch(r"[A-Za-z0-9_-]+", value):
        raise AgentProtocolError("invalid_encoding", "expected base64url without padding")
    return _base64url_decode(value)


def _valid_event_hash_bytes(event_hash: bytes) -> bytes:
    if len(event_hash) != 32:
        raise AgentProtocolError("invalid_event_hash", f"event hash must be 32 bytes, got {len(event_hash)}")
    return event_hash

File: packages/agent-protocols/src/identity.ts
import canonicalize from "canonicalize";
import { createHash } from "node:crypto";
import nacl from "tweetnacl";

import { protocolError } from "./errors.js";

export const AGENT_ID_PREFIX = "did:agent:";
export const DEFAULT_LIVE_WRITE_WINDOW_MS = 300_000;
export const DEFAULT_NONCE_TTL_MS = 300_000;
export const DEFAULT_REQUEST_JWT_TTL_SECS = 300;
export const MAX_NONCE_HEADER = "Max-Seen-Nonce";
export const MAX_SAFE_NONCE = Number.MAX_SAFE_INTEGER;

export type AgentId = string;

export interface Event<P = unknown> {
  protocol: string;
  type: string;
  actor: AgentId;
  created_at: number;
  nonce: number;
  room_id?: string;
  base_seq?: number;
  base_hash?: string;
  mentions?: AgentId[];
  payload: P;
  [key: string]: unknown;
}

export interface Envelope<P = unknown> {
  hash: string;
  event: Event<P>;
  signature: string;
}

export interface NonceRecord {
  maxNonce: number;
  expiresAt: number;
}

export interface NonceStore {
  checkAndUpdate(
    actor: AgentId,
    nonce: number,
    nowMs: number,
    ttlMs: number,
  ): number;
  maxNonce(actor: AgentId, nowMs: number): number | undefined;
}

export interface LiveWriteOptions {
  nowMs?: number;
  windowMs?: number;
  nonceTtlMs?: number;
}

export interface RequestJwtHeader {
  alg: "EdDSA";
  typ: "JWT";
  kid: AgentId;
}

export interface RequestBinding {
  /** The origin of the receiving service, e.g. `https://api.example.com`. */
  audience: string;
}

export interface RequestJwtClaims {
  iss: AgentId;
  sub: AgentId;
  /**
   * Origin of the receiving service: scheme, host, and non-default port with
   * no path. All Agent Protocols endpoints on one origin share this value.
   */
  aud: string;
  iat: number;
  exp: number;
}

export interface RequestAuthContext extends RequestBinding {
  nowSecs?: number;
  maxTtlSecs?: number;
}

export class AgentSigner {
  private constructor(private readonly keyPair: nacl.SignKeyPair) {}

  static generate(): AgentSigner {
    return new AgentSigner(nacl.sign.keyPair());
  }

  static fromSeed(seed: Uint8Array): AgentSigner {
    if (seed.byteLength !== 32) {
      throw protocolError(
        "invalid_private_key",
        `seed must be 32 bytes, got ${seed.byteLength}`,
      );
    }
    return new AgentSigner(nacl.sign.keyPair.fromSeed(seed));
  }

  agentId(): AgentId {
    return agentIdFromPublicKey(this.keyPair.publicKey);
  }

  publicKey(): Uint8Array {
    return new Uint8Array(this.keyPair.publicKey);
  }

  signEvent<P>(event: Event<P>): Envelope<P> {
    const hashBytes = eventHashBytes(event);
    return {
      hash: base64UrlEncode(hashBytes),
      event,
      signature: signEventHash(this.keyPair.secretKey, hashBytes),
    };
  }

  signRequestJwt(claims: RequestJwtClaims): string {
    const agentId = this.agentId();
    if (claims.iss !== agentId || claims.sub !== agentId) {
      throw protocolError(
        "invalid_jwt_claim",
        "iss and sub must match the signing agent id",
      );
    }

    const header: RequestJwtHeader = { alg: "EdDSA", typ: "JWT", kid: agentId };
    const encodedHeader = base64UrlEncode(
      new TextEncoder().encode(JSON.stringify(header)),
    );
    const encodedPayload = base64UrlEncode(
      new TextEncoder().encode(JSON.stringify(claims)),
    );
    const signingInput = `${encodedHeader}.${encodedPayload}`;
    const signature = nacl.sign.detached(
      new TextEncoder().encode(signingInput),
      this.keyPair.secretKey,
    );
    return `${signingInput}.${base64UrlEncode(signature)}`;
  }
}

export class MemoryNonceStore implements NonceStore {
  private readonly records = new Map<AgentId, NonceRecord>();

  checkAndUpdate(
    actor: AgentId,
    nonce: number,
    nowMs: number,
    ttlMs: number,
  ): number {
    validateNonce(nonce);
    if (ttlMs < 0) {
      throw protocolError(
        "invalid_nonce",
        "nonce cache ttl must be non-negative",
      );
    }
    const record = this.records.get(actor);
    if (record && record.expiresAt > nowMs && nonce <= record.maxNonce) {
      // Services rejecting for this reason MUST return the effective maximum
      // in the `Max-Seen-Nonce` response header; `data.max_nonce` carries it.
      throw protocolError(
        "nonce_not_greater",
        `nonce must be greater than accepted max nonce ${record.maxNonce}`,
        { max_nonce: record.maxNonce },
      );
    }
    this.records.set(actor, { maxNonce: nonce, expiresAt: nowMs + ttlMs });
    return nonce;
  }

  maxNonce(actor: AgentId, nowMs: number): number | undefined {
    const record = this.records.get(actor);
    return record && record.expiresAt > nowMs ? record.maxNonce : undefined;
  }
}

export class ClientNonceManager {
  constructor(private nextNonceValue = 1) {
    validateNonce(nextNonceValue);
  }

  peek(): number {
    return this.nextNonceValue;
  }

  nextNonce(): number {
    const nonce = this.nextNonceValue;
    validateNonce(nonce);
    this.nextNonceValue = nonce + 1;
    return nonce;
  }

  observeMaxNonce(maxNonce: number | string | null | undefined): void {
    if (maxNonce === null || maxNonce === undefined || maxNonce === "") return;
    const parsed = typeof maxNonce === "string" ? Number(maxNonce) : maxNonce;
    validateNonce(parsed);
    if (parsed >= this.nextNonceValue) {
      this.nextNonceValue = parsed + 1;
    }
  }
}

export function createEvent<P>(
  protocol: string,
  type: string,
  actor: AgentId,
  createdAt: number,
  nonce: number,
  payload: P,
): Event<P> {
  validateAgentId(actor);
  validateNonce(nonce);
  return {
    protocol,
    type,
    actor,
    created_at: createdAt,
    nonce,
    payload,
  };
}

export function withRoomId<P>(event: Event<P>, roomId: string): Event<P> {
  return {
    ...event,
    room_id: roomId,
  };
}

export function withRoomHead<P>(
  event: Event<P>,
  baseSeq: number,
  baseHash: string,
): Event<P> {
  validateNonce(baseSeq);
  if (baseHash.trim() === "") {
    throw protocolError("invalid_event", "base_hash must not be empty");
  }
  return {
    ...event,
    base_seq: baseSeq,
    base_hash: baseHash,
  };
}

export function withMentions<P>(
  event: Event<P>,
  mentions: AgentId[],
): Event<P> {
  for (const mention of mentions) validateAgentId(mention);
  return {
    ...event,
    mentions: [...mentions],
  };
}

export function withMention<P>(event: Event<P>, agentId: AgentId): Event<P> {
  validateAgentId(agentId);
  return {
    ...event,
    mentions: [...(event.mentions ?? []), agentId],
  };
}

export function agentIdFromPublicKey(publicKey: Uint8Array): AgentId {
  if (publicKey.byteLength !== 32) {
    throw protocolError(
      "invalid_public_key",
      `public key must be 32 bytes, got ${publicKey.byteLength}`,
    );
  }
  return `${AGENT_ID_PREFIX}${base64UrlEncode(publicKey)}`;
}

export function publicKeyBytes(agentId: AgentId): Uint8Array {
  if (typeof agentId !== "string") {
    throw protocolError("invalid_agent_id", "agent id must be a string");
  }
  const encoded = agentId.startsWith(AGENT_ID_PREFIX)
    ? agentId.slice(AGENT_ID_PREFIX.length)
    : undefined;
  if (!encoded) {
    throw protocolError(
      "invalid_agent_id",
      "agent id must start with did:agent:",
    );
  }
  const bytes = base64UrlDecode(encoded);
  if (bytes.byteLength !== 32) {
    throw protocolError(
      "invalid_public_key",
      `agent id public key must be 32 bytes, got ${bytes.byteLength}`,
    );
  }
  return bytes;
}

export function validateAgentId(agentId: AgentId): AgentId {
  publicKeyBytes(agentId);
  return agentId;
}

export function canonicalEventBytes(event: Event<unknown>): Uint8Array {
  const canonical = canonicalize(event);
  if (canonical === undefined) {
    throw protocolError(
      "canonical_json",
      "event cannot be represented as canonical JSON",
    );
  }
  return new TextEncoder().encode(canonical);
}

export function eventHash(event: Event<unknown>): string {
  return base64UrlEncode(eventHashBytes(event));
}

export function eventHashBytes(event: Event<unknown>): Uint8Array {
  validateNonce(event.nonce);
  return new Uint8Array(
    createHash("sha3-256")
      .update(canonicalEventBytes(event))
      .digest(),
  );
}

export function signEvent(
  secretKey: Uint8Array,
  event: Event<unknown>,
): string {
  return signEventHash(secretKey, eventHashBytes(event));
}

/**
 * Signs a precomputed 32-byte event hash. Signing a digest supplied by
 * another component without seeing the event it commits to (blind signing) is
 * NOT RECOMMENDED: `actor` and all event content are inside the digest, so a
 * blind signer can be tricked into signing arbitrary events attributed to its
 * key. Prefer `signEvent`, which canonicalizes and hashes the event itself.
 */
export function signEventHash(
  secretKey: Uint8Array,
  eventHash: Uint8Array,
): string {
  if (secretKey.byteLength !== 64) {
    throw protocolError(
      "invalid_private_key",
      `secret key must be 64 bytes, got ${secretKey.byteLength}`,
    );
  }
  return base64UrlEncode(
    nacl.sign.detached(validEventHashBytes(eventHash), secretKey),
  );
}

export function verifyEventHash(envelope: Envelope<unknown>): void {
  const expected = eventHash(envelope.event);
  if (expected !== envelope.hash) {
    throw protocolError(
      "invalid_event_hash",
      `invalid event hash: expected ${expected}, got ${envelope.hash}`,
    );
  }
}

export function verifySignature(envelope: Envelope<unknown>): void {
  verifyEventHashSignature(
    publicKeyBytes(envelope.event.actor),
    eventHashBytes(envelope.event),
    envelope.signature,
  );
}

export function verifyEventHashSignature(
  publicKey: Uint8Array,
  eventHash: Uint8Array,
  encodedSignature: string,
): void {
  if (publicKey.byteLength !== 32) {
    throw protocolError(
      "invalid_public_key",
      `public key must be 32 bytes, got ${publicKey.byteLength}`,
    );
  }
  const signature = base64UrlDecode(encodedSignature);
  if (signature.byteLength !== 64) {
    throw protocolError(
      "invalid_signature",
      `signature must be 64 bytes, got ${signature.byteLength}`,
    );
  }
  const ok = nacl.sign.detached.verify(
    validEventHashBytes(eventHash),
    signature,
    publicKey,
  );
  if (!ok) {
    throw protocolError("invalid_signature", "signature verification failed");
  }
}

export function verifyEnvelope(envelope: Envelope<unknown>): void {
  verifyEventHash(envelope);
  verifySignature(envelope);
}

export function verifyTimestamp(
  createdAt: number,
  nowMs: number,
  windowMs: number,
): void {
  if (windowMs < 0 || Math.abs(createdAt - nowMs) > windowMs) {
    throw protocolError(
      "timestamp_out_of_window",
      "timestamp is outside the allowed live-write window",
    );
  }
}

export function verifyLiveEnvelope(
  envelope: Envelope<unknown>,
  nonceStore: NonceStore,
  options: LiveWriteOptions = {},
): number {
  const nowMs = options.nowMs ?? unixTimeMillis();
  verifyEnvelope(envelope);
  verifyTimestamp(
    envelope.event.created_at,
    nowMs,
    options.windowMs ?? DEFAULT_LIVE_WRITE_WINDOW_MS,
  );
  return nonceStore.checkAndUpdate(
    envelope.event.actor,
    envelope.event.nonce,
    nowMs,
    options.nonceTtlMs ?? DEFAULT_NONCE_TTL_MS,
  );
}

export function createRequestBinding(audience: string): RequestBinding {
  return {
    audience,
  };
}

/**
 * Derives the request JWT `aud` from a request URL: the service origin —
 * scheme, host, and non-default port, with no path (Agent Identity Section 8).
 */
export function serviceOrigin(url: string): string {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw protocolError("invalid_url", `not a valid URL: ${url}`);
  }
  if (parsed.protocol !== "https:" && parsed.protocol !== "http:") {
    throw protocolError("invalid_url", `not an HTTP(S) URL: ${url}`);
  }
  return parsed.origin;
}

export function createRequestJwtClaims(
  agentId: AgentId,
  binding: RequestBinding,
  issuedAt: number,
  ttlSecs: number,
): RequestJwtClaims {
  return {
    iss: agentId,
    sub: agentId,
    aud: binding.audience,
    iat: issuedAt,
    exp: issuedAt + ttlSecs,
  };
}

export function verifyRequestJwt(
  token: string,
  context: RequestAuthContext,
): RequestJwtClaims {
  const parts = token.split(".");
  if (parts.length !== 3) {
    throw protocolError("invalid_jwt", "expected three compact JWS parts");
  }

  const header = JSON.parse(
    new TextDecoder().decode(base64UrlDecode(parts[0])),
  ) as RequestJwtHeader;
  const claims = JSON.parse(
    new TextDecoder().decode(base64UrlDecode(parts[1])),
  ) as RequestJwtClaims;
  const signature = base64UrlDecode(parts[2]);
  const signingInput = `${parts[0]}.${parts[1]}`;

  if (header.alg !== "EdDSA") {
    throw protocolError("invalid_jwt_claim", "alg must be EdDSA");
  }
  if (header.typ !== "JWT") {
    throw protocolError("invalid_jwt_claim", "typ must be JWT");
  }
  if (header.kid !== claims.iss || claims.iss !== claims.sub) {
    throw protocolError(
      "invalid_jwt_claim",
      "kid, iss, and sub must identify the same Agent ID",
    );
  }
  if (
    !nacl.sign.detached.verify(
      new TextEncoder().encode(signingInput),
      signature,
      publicKeyBytes(header.kid),
    )
  ) {
    throw protocolError(
      "invalid_signature",
      "JWT signature verification failed",
    );
  }

  if (claims.aud !== context.audience)
    throw protocolError("invalid_jwt_claim", "aud mismatch");

  const nowSecs = context.nowSecs ?? unixTimeSecs();
  const maxTtlSecs = context.maxTtlSecs ?? DEFAULT_REQUEST_JWT_TTL_SECS;
  if (claims.exp <= claims.iat) {
    throw protocolError("invalid_jwt_claim", "exp must be greater than iat");
  }
  if (claims.iat > nowSecs || claims.exp < nowSecs) {
    throw protocolError(
      "invalid_jwt_claim",
      "iat/exp outside valid time window",
    );
  }
  if (claims.exp - claims.iat > maxTtlSecs) {
    throw protocolError("invalid_jwt_claim", "JWT ttl exceeds maximum");
  }

  return claims;
}

export function unixTimeMillis(): number {
  return Date.now();
}

export function unixTimeSecs(): number {
  return Math.floor(Date.now() / 1000);
}

export function validateNonce(nonce: number): void {
  if (!Number.isSafeInteger(nonce) || nonce < 1 || nonce > MAX_SAFE_NONCE) {
    throw protocolError(
      "invalid_nonce",
      "nonce must be a positive safe integer",
    );
  }
}

function base64UrlEncode(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64url");
}

/**
 * Canonical base64url decoding: URL-safe alphabet, no padding, zero trailing
 * bits. Receivers MUST reject non-canonical encodings, otherwise one value
 * gains multiple distinct string forms and corrupts string-keyed comparisons.
 */
function base64UrlDecode(value: string): Uint8Array {
  const bytes = new Uint8Array(Buffer.from(value, "base64url"));
  if (Buffer.from(bytes).toString("base64url") !== value) {
    throw protocolError(
      "invalid_encoding",
      "expected canonical base64url without padding",
    );
  }
  return bytes;
}

function validEventHashBytes(eventHash: Uint8Array): Uint8Array {
  if (eventHash.byteLength !== 32) {
    throw protocolError(
      "invalid_event_hash",
      `event hash must be 32 bytes, got ${eventHash.byteLength}`,
    );
  }
  return eventHash;
}
